From 107684cf3b9997d4e9084dc87259a0841b48946e Mon Sep 17 00:00:00 2001
From: Chen Meng J <meng.j.chen@intel.com>
Date: Wed, 31 Aug 2022 11:10:13 +0800
Subject: [PATCH 003/102] revert: media: Add poll support

Signed-off-by: Chen Meng J <meng.j.chen@intel.com>
Signed-off-by: zouxiaoh <xiaohong.zou@intel.com>
---
 drivers/media/mc/mc-device.c | 34 ----------------------------------
 1 file changed, 34 deletions(-)

diff --git a/drivers/media/mc/mc-device.c b/drivers/media/mc/mc-device.c
index 127cbb542b59..106ce313ae2b 100644
--- a/drivers/media/mc/mc-device.c
+++ b/drivers/media/mc/mc-device.c
@@ -501,39 +501,6 @@ out_free:
 	return ret;
 }
 
-static unsigned int media_device_poll(struct file *filp,
-					struct poll_table_struct *wait)
-{
-	/*
-	struct media_device_fh *fh = media_device_fh(filp);
-	struct media_device *mdev = fh->fh.devnode->media_dev;
-	unsigned int poll_events = poll_requested_events(wait);
-	int ret = 0;
-
-	if (poll_events & (POLLIN | POLLOUT))
-		return POLLERR;
-
-	if (poll_events & POLLPRI) {
-		unsigned long flags;
-		bool empty;
-
-		spin_lock_irqsave(&mdev->req_lock, flags);
-		empty = list_empty(&fh->kevents.head);
-		spin_unlock_irqrestore(&mdev->req_lock, flags);
-
-		if (empty)
-			poll_wait(filp, &fh->kevents.wait, wait);
-		else
-			ret |= POLLPRI;
-	}
-
-	return ret;
-	*/
-
-	printk("media_device_poll func\n");
-	return 0;
-}
-
 #ifdef CONFIG_COMPAT
 
 struct media_links_enum32 {
@@ -599,7 +566,6 @@ static const struct media_file_operations media_device_fops = {
 	.owner = THIS_MODULE,
 	.open = media_device_open,
 	.ioctl = media_device_ioctl,
-	.poll = media_device_poll,
 #ifdef CONFIG_COMPAT
 	.compat_ioctl = media_device_compat_ioctl,
 #endif /* CONFIG_COMPAT */
-- 
2.17.1

