From 171e440653fa1bf288f9fbaccb0c822e89cb6888 Mon Sep 17 00:00:00 2001
From: Chen Meng J <meng.j.chen@intel.com>
Date: Wed, 31 Aug 2022 14:29:14 +0800
Subject: [PATCH 013/102] revert: media: v4l: Add new vectorised pixel formats

Signed-off-by: Chen Meng J <meng.j.chen@intel.com>
Signed-off-by: zouxiaoh <xiaohong.zou@intel.com>
---
 include/uapi/linux/videodev2.h | 31 -------------------------------
 1 file changed, 31 deletions(-)

diff --git a/include/uapi/linux/videodev2.h b/include/uapi/linux/videodev2.h
index 1bbe0d83d9fe..9260791b8438 100644
--- a/include/uapi/linux/videodev2.h
+++ b/include/uapi/linux/videodev2.h
@@ -593,7 +593,6 @@ struct v4l2_pix_format {
 #define V4L2_PIX_FMT_VUYA32  v4l2_fourcc('V', 'U', 'Y', 'A') /* 32  VUYA-8-8-8-8  */
 #define V4L2_PIX_FMT_VUYX32  v4l2_fourcc('V', 'U', 'Y', 'X') /* 32  VUYX-8-8-8-8  */
 #define V4L2_PIX_FMT_M420    v4l2_fourcc('M', '4', '2', '0') /* 12  YUV 4:2:0 2 lines y, 1 line uv interleaved */
-#define V4L2_PIX_FMT_Y210    v4l2_fourcc('Y', '2', '1', '0') /* 20  YUV 4:2:2 10-bit yuyv*/
 
 /* two planes -- one Y, one Cr + Cb interleaved  */
 #define V4L2_PIX_FMT_NV12    v4l2_fourcc('N', 'V', '1', '2') /* 12  Y/CbCr 4:2:0  */
@@ -675,36 +674,6 @@ struct v4l2_pix_format {
 #define V4L2_PIX_FMT_SGRBG16 v4l2_fourcc('G', 'R', '1', '6') /* 16  GRGR.. BGBG.. */
 #define V4L2_PIX_FMT_SRGGB16 v4l2_fourcc('R', 'G', '1', '6') /* 16  RGRG.. GBGB.. */
 
-
-/* Raw bayer vector formats. */
-#define V4L2_PIX_FMT_SBGGR8_16V32      v4l2_fourcc('b', 'V', '0', 'A')
-#define V4L2_PIX_FMT_SGBRG8_16V32      v4l2_fourcc('b', 'V', '0', 'B')
-#define V4L2_PIX_FMT_SGRBG8_16V32      v4l2_fourcc('b', 'V', '0', 'C')
-#define V4L2_PIX_FMT_SRGGB8_16V32      v4l2_fourcc('b', 'V', '0', 'D')
-#define V4L2_PIX_FMT_SBGGR10V32        v4l2_fourcc('b', 'V', '0', 'E')
-#define V4L2_PIX_FMT_SGBRG10V32        v4l2_fourcc('b', 'V', '0', 'F')
-#define V4L2_PIX_FMT_SGRBG10V32        v4l2_fourcc('b', 'V', '0', 'G')
-#define V4L2_PIX_FMT_SRGGB10V32        v4l2_fourcc('b', 'V', '0', 'H')
-#define V4L2_PIX_FMT_SBGGR12V32        v4l2_fourcc('b', 'V', '0', 'I')
-#define V4L2_PIX_FMT_SGBRG12V32        v4l2_fourcc('b', 'V', '0', 'J')
-#define V4L2_PIX_FMT_SGRBG12V32        v4l2_fourcc('b', 'V', '0', 'K')
-#define V4L2_PIX_FMT_SRGGB12V32        v4l2_fourcc('b', 'V', '0', 'L')
-#define V4L2_PIX_FMT_SBGGR14V32        v4l2_fourcc('b', 'V', '0', 'M')
-#define V4L2_PIX_FMT_SGBRG14V32        v4l2_fourcc('b', 'V', '0', 'N')
-#define V4L2_PIX_FMT_SGRBG14V32        v4l2_fourcc('b', 'V', '0', 'O')
-#define V4L2_PIX_FMT_SRGGB14V32        v4l2_fourcc('b', 'V', '0', 'P')
-
-/* BEGIN remove once the user space has been updated */
-#define V4L2_PIX_FMT_SBGGR8V32 v4l2_fourcc('b', 'V', '0', 'A')
-#define V4L2_PIX_FMT_SGBRG8V32 v4l2_fourcc('b', 'V', '0', 'B')
-#define V4L2_PIX_FMT_SGRBG8V32 v4l2_fourcc('b', 'V', '0', 'C')
-#define V4L2_PIX_FMT_SRGGB8V32 v4l2_fourcc('b', 'V', '0', 'D')
-/* END remove once the user space has been updated */
-
-/* YUV vector formats. */
-#define V4L2_PIX_FMT_UYVY_V32          v4l2_fourcc('y', 'V', '3', '2')
-#define V4L2_PIX_FMT_YUYV420_V32       v4l2_fourcc('y', '0', '3', '2')
-
 /* HSV formats */
 #define V4L2_PIX_FMT_HSV24 v4l2_fourcc('H', 'S', 'V', '3')
 #define V4L2_PIX_FMT_HSV32 v4l2_fourcc('H', 'S', 'V', '4')
-- 
2.17.1

