diff --git a/drivers/platform/x86/intel/int3472/clk_and_regulator.c b/drivers/platform/x86/intel/int3472/clk_and_regulator.c
index fd4df039093f..8cdffb13a44b 100644
--- a/drivers/platform/x86/intel/int3472/clk_and_regulator.c
+++ b/drivers/platform/x86/intel/int3472/clk_and_regulator.c
@@ -247,8 +247,11 @@ int skl_int3472_register_regulator(struct int3472_discrete_device *int3472,
 	int3472->regulator.gpio = acpi_get_and_request_gpiod(path, agpio->pin_table[0],
 							     "int3472,regulator");
 	if (IS_ERR(int3472->regulator.gpio)) {
-		dev_err(int3472->dev, "Failed to get regulator GPIO line\n");
-		return PTR_ERR(int3472->regulator.gpio);
+		ret = (int)PTR_ERR(int3472->regulator.gpio);
+		int3472->regulator.gpio = NULL;
+		if (ret != -EPROBE_DEFER)
+			dev_err(int3472->dev, "Failed to get regulator GPIO line: %d", ret);
+		return ret;
 	}
 
 	/* Ensure the pin is in output mode and non-active state */

