From adf7cdb02005e1d0875396679b2e1fb5663d1b31 Mon Sep 17 00:00:00 2001
From: Dongcheng Yan <dongcheng.yan@intel.com>
Date: Tue, 30 Jul 2024 16:50:07 +0800
Subject: [PATCH] upstream: Use module parameter to set psys freq

Signed-off-by: Dongcheng Yan <dongcheng.yan@intel.com>
Signed-off-by: zouxiaoh <xiaohong.zou@intel.com>
---
 drivers/media/pci/intel/ipu6/ipu6.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/media/pci/intel/ipu6/ipu6.c b/drivers/media/pci/intel/ipu6/ipu6.c
index f6d42a432d9f..4bb270d9fe2a 100644
--- a/drivers/media/pci/intel/ipu6/ipu6.c
+++ b/drivers/media/pci/intel/ipu6/ipu6.c
@@ -37,6 +37,10 @@ static unsigned int isys_freq_override;
 module_param(isys_freq_override, uint, 0660);
 MODULE_PARM_DESC(isys_freq_override, "Override ISYS freq(mhz)");

+static unsigned int psys_freq_override;
+module_param(psys_freq_override, uint, 0660);
+MODULE_PARM_DESC(psys_freq_override, "Override PSYS freq(mhz)");
+
 #define IPU6_PCI_BAR		0

 struct ipu6_cell_program {
@@ -445,6 +449,15 @@ ipu6_psys_init(struct pci_dev *pdev, struct device *parent,
 	pdata->base = base;
 	pdata->ipdata = ipdata;

+	/* Override the isys freq */
+	if (psys_freq_override >= BUTTRESS_MIN_FORCE_PS_FREQ &&
+	    psys_freq_override <= BUTTRESS_MAX_FORCE_PS_FREQ) {
+		ctrl->ratio = psys_freq_override / BUTTRESS_PS_FREQ_STEP;
+		ctrl->qos_floor = psys_freq_override;
+		dev_dbg(&pdev->dev, "Override the psys freq:%u(mhz)\n",
+			psys_freq_override);
+	}
+
 	psys_adev = ipu6_bus_initialize_device(pdev, parent, pdata, ctrl,
 					       IPU6_PSYS_NAME);
 	if (IS_ERR(psys_adev)) {
--
2.34.1

