From e124e204dc247512666675066c252330e425daa6 Mon Sep 17 00:00:00 2001
From: Chen Meng J <meng.j.chen@intel.com>
Date: Wed, 31 Aug 2022 10:50:47 +0800
Subject: [PATCH 002/102] revert: v4l2-mc: fix entity use_count issue for dual
 sensors

Signed-off-by: Chen Meng J <meng.j.chen@intel.com>
Signed-off-by: zouxiaoh <xiaohong.zou@intel.com>
---
 drivers/media/v4l2-core/v4l2-mc.c | 12 ++++--------
 1 file changed, 4 insertions(+), 8 deletions(-)

diff --git a/drivers/media/v4l2-core/v4l2-mc.c b/drivers/media/v4l2-core/v4l2-mc.c
index 63263888ecdf..82f7646b2b01 100644
--- a/drivers/media/v4l2-core/v4l2-mc.c
+++ b/drivers/media/v4l2-core/v4l2-mc.c
@@ -431,13 +431,12 @@ EXPORT_SYMBOL_GPL(v4l2_create_fwnode_links);
  *
  * Return the total number of users of all video device nodes in the pipeline.
  */
-static int pipeline_pm_use_count(struct media_pad *pad,
+static int pipeline_pm_use_count(struct media_entity *entity,
 	struct media_graph *graph)
 {
-	struct media_entity *entity = pad->entity;
 	int use = 0;
 
-	media_graph_walk_start(graph, pad);
+	media_graph_walk_start(graph, &entity->pads[0]);
 
 	while ((entity = media_graph_walk_next(graph))) {
 		if (is_media_entity_v4l2_video_device(entity))
@@ -564,17 +563,14 @@ int v4l2_pipeline_link_notify(struct media_link *link, u32 flags,
 	int sink_use;
 	int ret = 0;
 
-	source_use = pipeline_pm_use_count(link->source, graph);
-	sink_use = pipeline_pm_use_count(link->sink, graph);
+	source_use = pipeline_pm_use_count(source, graph);
+	sink_use = pipeline_pm_use_count(sink, graph);
 
 	if (notification == MEDIA_DEV_NOTIFY_POST_LINK_CH &&
 	    !(flags & MEDIA_LNK_FL_ENABLED)) {
 		/* Powering off entities is assumed to never fail. */
 		pipeline_pm_power(source, -sink_use, graph);
 		pipeline_pm_power(sink, -source_use, graph);
-
-		source->use_count = 0;
-		sink->use_count = 0;
 		return 0;
 	}
 
-- 
2.17.1

